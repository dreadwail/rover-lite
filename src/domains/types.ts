import { Action, AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

import { RoverSearchState } from '../state/types';

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export type GenericThunkAction<T, S extends RoverSearchState = RoverSearchState> = ThunkAction<
  T,
  S,
  {},
  AnyAction
>;
export type PromiseThunkAction<
  T,
  S extends RoverSearchState = RoverSearchState
> = GenericThunkAction<Promise<T>, S>;

export type MapDispatchToProps<
  TDispatchProps,
  TOwnProps = {},
  TState extends RoverSearchState = RoverSearchState,
  A extends Action = Action
> = <E>(dispatch: ThunkDispatch<TState, E, A>, ownProps: TOwnProps) => TDispatchProps;
