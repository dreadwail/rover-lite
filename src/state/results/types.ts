export interface Badge {
  readonly content: string;
  readonly imageUrl: string;
  readonly title: string;
}

export interface Listing {
  readonly thumbnailUrl: string;
  readonly rank: number;
  readonly name: string;
  readonly badges: Badge[];
  readonly tagline: string;
  readonly neighborhood: string;
  readonly city: string;
  readonly state: string;
  readonly zip: string;
  readonly repeatClientCount: number;
  readonly reviewCount: number;
  readonly ratingAverage: number;
  readonly featuredReview: string;
  readonly currencySymbol: string;
  readonly price: number;
  readonly priceUnit: string;
  readonly webUrl: string;
}

export interface Results {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly isStale: boolean;
  readonly listings: Listing[];
}
