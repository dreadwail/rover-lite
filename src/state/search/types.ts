import { SearchType } from '../searchTypes/types';

export type SearchInputs = Record<string, any>;

export interface Search {
  readonly searchType: SearchType;
  readonly latitude: number;
  readonly longitude: number;
  readonly inputs: SearchInputs;
}
