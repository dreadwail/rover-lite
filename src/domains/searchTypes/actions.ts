import { Action } from 'redux';

import { request } from '../../api';
import { Endpoints } from '../../state/searchTypes/types';
import { PromiseThunkAction } from '../types';

export interface FetchSearchTypesPendingAction extends Action {
  readonly type: 'FETCH_SEARCH_TYPES.PENDING';
}

export interface FetchSearchTypesFulfilledAction extends Action {
  readonly type: 'FETCH_SEARCH_TYPES.FULFILLED';
  readonly payload: Endpoints;
}

export interface FetchSearchTypesRejectedAction extends Action {
  readonly type: 'FETCH_SEARCH_TYPES.REJECTED';
  readonly payload: {
    readonly error: any;
  };
  readonly error: true;
}

const fetchSearchTypesPending = (): FetchSearchTypesPendingAction => ({
  type: 'FETCH_SEARCH_TYPES.PENDING',
});

const fetchSearchTypesFulfilled = (endpoints: Endpoints): FetchSearchTypesFulfilledAction => ({
  type: 'FETCH_SEARCH_TYPES.FULFILLED',
  payload: endpoints,
});

const fetchSearchTypesRejected = (error: any): FetchSearchTypesRejectedAction => ({
  type: 'FETCH_SEARCH_TYPES.REJECTED',
  payload: {
    error,
  },
  error: true,
});

export type FetchSearchTypes = () => PromiseThunkAction<void>;
export const fetchSearchTypes: FetchSearchTypes = () => async dispatch => {
  dispatch(fetchSearchTypesPending());
  try {
    const response = await request<Endpoints>(
      'GET',
      'https://www.rover.com/api/v3/search-endpoints/'
    );
    dispatch(fetchSearchTypesFulfilled(response));
  } catch (err) {
    dispatch(fetchSearchTypesRejected(err));
  }
};
