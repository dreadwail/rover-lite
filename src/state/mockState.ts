import { RoverSearchState } from './types';

type Primitive = boolean | number | string;

export type RecursivePartial<ThisLevel> = {
  [Key in keyof ThisLevel]?: ThisLevel[Key] extends Primitive
    ? ThisLevel[Key]
    : RecursivePartial<ThisLevel[Key]>
};

export const createMockState = (
  interestingStateKeys?: RecursivePartial<RoverSearchState>
): RoverSearchState => (interestingStateKeys as unknown) as RoverSearchState;
