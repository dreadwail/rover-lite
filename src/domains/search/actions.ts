import { Action } from 'redux';

import { SearchType } from '../../state/searchTypes/types';
import { GenericThunkAction } from '../types';

import { getInputs } from './selectors';

export interface SearchTypeChangedAction extends Action {
  readonly type: 'SEARCH.TYPE_CHANGED';
  readonly payload: SearchType;
}

export interface SearchFieldChangedAction extends Action {
  readonly type: 'SEARCH.FIELD_CHANGED';
  readonly payload: {
    readonly key: string;
    readonly value: any;
  };
}

export const searchTypeChanged = (searchType: SearchType): SearchTypeChangedAction => ({
  type: 'SEARCH.TYPE_CHANGED',
  payload: searchType,
});

export type SearchFieldChanged = (key: string, value: any) => GenericThunkAction<void>;
export const searchFieldChanged: SearchFieldChanged = (key, value) => async (
  dispatch,
  getState
) => {
  const state = getState();
  const inputs = getInputs(state);
  const startDate = inputs.start_date;
  const endDate = inputs.end_date;
  if (startDate != null && key === 'end_date' && startDate > value) {
    return;
  }
  if (endDate != null && key === 'start_date' && value > endDate) {
    return;
  }
  const action: SearchFieldChangedAction = {
    type: 'SEARCH.FIELD_CHANGED',
    payload: { key, value },
  };
  dispatch(action);
};
