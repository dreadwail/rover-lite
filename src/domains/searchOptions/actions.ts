import { Action } from 'redux';

import { request } from '../../api';
import { Options } from '../../state/searchOptions/types';
import { SearchType } from '../../state/searchTypes/types';
import { getEndpointUrl } from '../searchTypes/selectors';
import { PromiseThunkAction } from '../types';

import { normalize, OptionsResponse } from './normalize';

export interface FetchSearchOptionsPendingAction extends Action {
  readonly type: 'FETCH_SEARCH_OPTIONS.PENDING';
  readonly payload: {
    readonly searchType: SearchType;
  };
}

export interface FetchSearchOptionsFulfilledAction extends Action {
  readonly type: 'FETCH_SEARCH_OPTIONS.FULFILLED';
  readonly payload: {
    readonly searchType: SearchType;
    readonly options: Options;
  };
}

export interface FetchSearchOptionsRejectedAction extends Action {
  readonly type: 'FETCH_SEARCH_OPTIONS.REJECTED';
  readonly payload: {
    readonly searchType: SearchType;
    readonly error: any;
  };
  readonly error: true;
}

const fetchSearchOptionsPending = (searchType: SearchType): FetchSearchOptionsPendingAction => ({
  type: 'FETCH_SEARCH_OPTIONS.PENDING',
  payload: {
    searchType,
  },
});

const fetchSearchOptionsFulfilled = (
  searchType: SearchType,
  options: Options
): FetchSearchOptionsFulfilledAction => ({
  type: 'FETCH_SEARCH_OPTIONS.FULFILLED',
  payload: {
    searchType,
    options,
  },
});

const fetchSearchOptionsRejected = (
  searchType: SearchType,
  error: any
): FetchSearchOptionsRejectedAction => ({
  type: 'FETCH_SEARCH_OPTIONS.REJECTED',
  payload: {
    searchType,
    error,
  },
  error: true,
});

export type FetchSearchOptions = (searchType: SearchType) => PromiseThunkAction<void>;
export const fetchSearchOptions: FetchSearchOptions = searchType => async (dispatch, getState) => {
  dispatch(fetchSearchOptionsPending(searchType));
  try {
    const state = getState();
    const endpointUrl = getEndpointUrl(state, searchType);
    if (!endpointUrl) {
      throw new Error(`Unhandled search type: ${searchType}`);
    }
    const response = await request<OptionsResponse>('OPTIONS', endpointUrl);
    const normalizedOptions = normalize(response);
    dispatch(fetchSearchOptionsFulfilled(searchType, normalizedOptions));
  } catch (err) {
    dispatch(fetchSearchOptionsRejected(searchType, err));
  }
};
