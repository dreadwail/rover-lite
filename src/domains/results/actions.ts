import { Action } from 'redux';

import { request } from '../../api/http';
import { urlEncodeQuery } from '../../api/query';
import { Listing } from '../../state/results/types';
import { SearchInputs } from '../../state/search/types';
import debouncedAction from '../debouncedThunk';
import { getInputs, getLatLong, getSearchType } from '../search/selectors';
import { PromiseThunkAction } from '../types';

import { normalize, RawSearchResponse } from './normalize';

export interface FetchResultsPendingAction extends Action {
  readonly type: 'FETCH_RESULTS.PENDING';
}

export interface FetchResultsFulfilledAction extends Action {
  readonly type: 'FETCH_RESULTS.FULFILLED';
  readonly payload: Listing[];
}

export interface FetchResultsRejectedAction extends Action {
  readonly type: 'FETCH_RESULTS.REJECTED';
  readonly payload: {
    readonly error: any;
  };
  readonly error: true;
}

const fetchResultsPending = (): FetchResultsPendingAction => ({
  type: 'FETCH_RESULTS.PENDING',
});

const fetchResultsFulfilled = (listings: Listing[]): FetchResultsFulfilledAction => ({
  type: 'FETCH_RESULTS.FULFILLED',
  payload: listings,
});

const fetchResultsRejected = (error: any): FetchResultsRejectedAction => ({
  type: 'FETCH_RESULTS.REJECTED',
  payload: {
    error,
  },
  error: true,
});

export type FetchResults = () => PromiseThunkAction<void>;
export const fetchResults: FetchResults = () => async (dispatch, getState) => {
  dispatch(fetchResultsPending());
  try {
    const state = getState();

    const searchType = getSearchType(state);
    const [latitude, longitude] = getLatLong(state);

    const defaultInputs: SearchInputs = {
      centerlat: latitude,
      centerlng: longitude,
      service_type: searchType,
    };
    const userInputs = getInputs(state);
    const inputs = { ...defaultInputs, ...userInputs };
    const encodedQuery = urlEncodeQuery(inputs);

    const urlBase = 'https://www.rover.com/api/v3/search';
    const url = `${urlBase}/${searchType}/?${encodedQuery}`;

    const response = await request<RawSearchResponse>('GET', url);

    const normalizedListings = normalize(response);
    dispatch(fetchResultsFulfilled(normalizedListings));
  } catch (err) {
    dispatch(fetchResultsRejected(err));
  }
};

export const debouncedFetchResults = debouncedAction(fetchResults, 250);
