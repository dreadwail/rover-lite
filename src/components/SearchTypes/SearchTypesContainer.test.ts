jest.mock('../../domains/searchTypes/actions');
jest.mock('../../domains/search/actions');

import { Action, Dispatch } from 'redux';

import { searchTypeChanged } from '../../domains/search/actions';
import { fetchSearchTypes } from '../../domains/searchTypes/actions';
import { createMockState } from '../../state/mockState';
import { Search } from '../../state/search/types';
import { SearchType, SearchTypes } from '../../state/searchTypes/types';

import {
  mapDispatchToProps,
  mapStateToProps,
  MappedDispatchProps,
  MappedStateProps,
} from './SearchTypesContainer';

describe('SearchTypesContainer', () => {
  describe('.mapStateToProps', () => {
    let mappedStateProps: MappedStateProps;

    beforeEach(() => {
      const searchTypes: SearchTypes = {
        isFetching: false,
        hasFailed: false,
        endpoints: {
          'overnight-boarding': 'https://www.rover.com/api/v3/search/overnight-boarding/',
        },
      };
      const search: Search = {
        searchType: 'overnight-boarding',
        inputs: {},
      };
      const state = createMockState({ searchTypes, search });
      mappedStateProps = mapStateToProps(state, {});
    });

    it('returns the correctly mapped props', () => {
      expect(mappedStateProps).toMatchSnapshot();
    });
  });

  describe('.mapDispatchToProps', () => {
    let dispatch: Dispatch;
    let mappedDispatchProps: MappedDispatchProps;

    beforeEach(() => {
      dispatch = jest.fn().mockName('dispatch');
      mappedDispatchProps = mapDispatchToProps(dispatch, {});
    });

    describe('.fetchSearchTypes', () => {
      let fetchSearchTypesAction: Action;

      beforeEach(() => {
        fetchSearchTypesAction = { type: 'TEST_ACTION' };
        (fetchSearchTypes as jest.Mock).mockReturnValue(fetchSearchTypesAction);
        mappedDispatchProps.fetchSearchTypes();
      });

      it('calls the correct action creator', () => {
        expect(fetchSearchTypes).toHaveBeenCalled();
      });

      it('dispatches the correct action', () => {
        expect(dispatch).toHaveBeenCalledWith(fetchSearchTypesAction);
      });
    });

    describe('.onSearchTypeSelected', () => {
      let searchTypeChangedAction: Action;

      const searchType: SearchType = 'overnight-boarding';

      beforeEach(() => {
        searchTypeChangedAction = { type: 'TEST_ACTION' };
        (searchTypeChanged as jest.Mock).mockReturnValue(searchTypeChangedAction);
        mappedDispatchProps.onSearchTypeSelected(searchType);
      });

      it('calls the correct action creator', () => {
        expect(searchTypeChanged).toHaveBeenCalledWith(searchType);
      });

      it('dispatches the correct action', () => {
        expect(dispatch).toHaveBeenCalledWith(searchTypeChangedAction);
      });
    });
  });
});
