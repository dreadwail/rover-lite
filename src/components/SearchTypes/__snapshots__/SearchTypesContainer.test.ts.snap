// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchTypesContainer .mapStateToProps returns the correctly mapped props 1`] = `
Object {
  "hasFailed": false,
  "isFetching": false,
  "searchType": "overnight-boarding",
  "searchTypes": Array [
    "overnight-boarding",
  ],
}
`;
